File: C#/MergeSort.cs
using System;
class MergeSort {
	void merge(int[] arr, int l, int m, int r)
	{
		int n1 = m - l + 1;
		int n2 = r - m;
		int[] L = new int[n1];
		int[] R = new int[n2];
		int i, j;

		for (i = 0; i < n1; ++i)
			L[i] = arr[l + i];
      
		for (j = 0; j < n2; ++j)
			R[j] = arr[m + 1 + j];

		i = 0;
		j = 0;
		int k = l;
    
		while (i < n1 && j < n2) {
			if (L[i] <= R[j]) {
				arr[k] = L[i];
				i++;
			}
			else {
				arr[k] = R[j];
				j++;
			}
			k++;
		}

		while (i < n1) {
			arr[k] = L[i];
			i++;
			k++;
		}

		while (j < n2) {
			arr[k] = R[j];
			j++;
			k++;
		}
	}

	void sort(int[] arr, int l, int r)
	{
		if (l < r) {
			int m = l + (r - l) / 2;
			sort(arr, l, m);
			sort(arr, m + 1, r);
			merge(arr, l, m, r);
		}
	}

	static void printArray(int[] arr)
	{
		int n = arr.Length;
		for (int i = 0; i < n; ++i)
			Console.Write(arr[i] + " ");
		Console.WriteLine();
	}

	public static void Main(String[] args)
	{
		int[] arr = { 24, 85, 46, 2, 64, 34 };
    
		Console.WriteLine("Inserted Array: ");
		printArray(arr);
    
		MergeSort ob = new MergeSort();
		ob.sort(arr, 0, arr.Length - 1);
    
		Console.WriteLine("\nSorted array: ");
		printArray(arr);
	}
}

File: Java/mergeSort.java
import java.util.*;

public class mergerSort {
    static int[] MergeSort(int[] arr) {
        if (arr.length == 1) {
            return arr;
        }
        int mid = arr.length / 2;
        int[] left = MergeSort(Arrays.copyOfRange(arr, 0, mid));
        int[] right = MergeSort(Arrays.copyOfRange(arr, mid, arr.length));
        return merge(left, right);
    }

    private static int[] merge(int[] first, int[] second) {
        int[] mix = new int[first.length + second.length];

        int i = 0;
        int j = 0;
        int k = 0;

        while (i < first.length && j < second.length) {
            if (first[i] < second[j]) {
                mix[k] = first[i];
                i++;
            } else {
                mix[k] = second[j];
                j++;
            }
            k++;
        }
        // it may be possible that one of the arrays is not complete
        // copy the remaining element
        while (i < first.length) {
            mix[k] = first[i];
            i++;
            k++;
        }
        while (j < second.length) {
            mix[k] = second[j];
            j++;
            k++;
        }
        return mix;
    }

    public static void main(String[] args) {
        int[] arr = { 5, 4, 3, 30, 244 };
        arr = MergeSort(arr);
        System.out.println(Arrays.toString(arr));
    }
}
